### PR TITLE
Add path icon for menubar

### PR DESCRIPTION
**Pull request by salorium**
Add "user path icon" for menubar

### CODE REVIEW

**Comment by Elv13** [comment]
Hello,

Thanks for this pull request. I would like to mention that the variable $XDG_DATA_DIRS should also be parsed to get the list of all supported XDG locations, not only / and ~/.

See https://developer.gnome.org/basedir-spec/

Please use the glib functions for this, not manually parsing the string in lua

https://developer.gnome.org/glib/stable/glib-Miscellaneous-Utility-Functions.html#g-get-system-data-dirs
https://developer.gnome.org/glib/stable/glib-Miscellaneous-Utility-Functions.html#g-get-user-special-dir

**Comment by salorium** [comment]
Hello,
I rewrite a code and i use a XDG_DATA_DIRS variable.

**Comment by Elv13** [comment]
Hello again,

As I proposed in the last comment, I would prefer to see an implementation that use the glib methods instead of yet another string split hardcoded in Lua. Awesome already use glib for most operations and it is implemented in C. It would be faster and wont add any new parsing code.

**Comment by salorium** [comment]
(google traduction)Ok, sorry i am a beginner and i dont speak english.
I do not know or do I add the code in awesome in c.
Sorry

**Comment by Elv13** [comment]
Looking at your commit log, I guess you speak french, so here I go:

Tu n'as pas besoin d'ajouter du code C pour faire cela. Awesome utilise un binding Lua pour GLIb qui s'appel LGI. Cela permet d'utiliser les fonction GLib dans le code lua. Il faut faire:

```lua
local glib  = require("lgi").GLib
```

Pour utiliser les fonctions, il faut enlever les prefix (ici: `glib`, parfois `cairo` ou `gio`) et appeler la fonction:

```lua
local glib  = require("lgi").GLib

for k,dossier in ipairs(glib.get_system_config_dirs()) do
  print("Un dossier:", dossier)
end
```
Dans ce cas, cela va printer `/etc/xdg`. Il existe d'autres fonctions pour les autre types de dossiers dans la page d'api glib que je t'ai linker au post precedent.

J'espere que ca va t'aider

**Comment by salorium** [comment]
Thank you,
Merci de ton aide surtout en Français. c'est super !!!
J'ai essayé de faire au mieux, j'espère que cela te conviendras.

**Comment by Elv13** [comment]
Thanks for the changes,

It is almost fine. There is three things I would like to see fixed. First, the identation is wrong. And then, the require() call must be at the beginning of the file, calling it in the function does work, but have a slight overhead. Finally, please replace "dossier" with "directory" or "dir".

---

C'est presque ok, il faut juste fixer l'indentation, changer "dossier" pour "dir" et mettre le require() en haut du fichier avec les autres.

**Comment by salorium** [comment]
Voila c'est fait, merci de m'avoir guidé.

File: lib/menubar/utils.lua.in [hunk 1]
```diff
@@ -80,12 +80,15 @@ function utils.lookup_icon(icon_file)
...
         local icon_theme = theme.icon_theme
-        if icon_theme then
-            table.insert(icon_theme_paths, '/usr/share/icons/' .. icon_theme .. '/')
-            -- TODO also look in parent icon themes, as in freedesktop.org specification
+        local paths = glib.get_system_data_dirs()
+        table.insert(paths,1,glib.get_user_data_dir())
+        for k,dir in ipairs(paths)do
+            if icon_theme then
+                table.insert(icon_theme_paths, dir..'/icons/' .. icon_theme .. '/')
+                -- TODO also look in parent icon themes, as in freedesktop.org specification
```
Comment (blueyed) [inline lib/menubar/utils.lua.in]: @salorium 
Isn't this TODO now obsolete / fixed?

Comment (salorium) [inline lib/menubar/utils.lua.in]: Yes, it's fixed

**Comment by salorium** [comment]
Ok, i remove todo

**Comment by Elv13** [comment]
Look good to me, can anybody else confirm it works for them (I use gtk to lookup for icons, it is much faster but add gtk as a dependency for awesome, so your solution is probably a better idea for most peoples). If nobody step up, I will merge this

**Comment by blueyed** [comment]
Would it make sense to not build the lookup path again and again, but only once?

**Comment by Elv13** [comment]
In my opinion, we should check for GTK with a pcall then delegate icon lookup. If that fail, then looping 3 directories wont make that much of a difference. The GTK code path is (afaik) cached and in C. I use it in my implementation and it is night and day when loading the full menu.

**Comment by blueyed** [comment]
Ok. For now I think we should take this PR and then improve upon it.

A new dependency is probably not that good.

I'll commit it to master later, amended and with fixed whitespace.

**Comment by salorium** [comment]
@blueyed 
I read your last post, and i change a code for build the lookup path only once.

**Comment by Elv13** [comment]
@blueyed The advantage is that we can loop for GIO, GvFS and GTK at runtime then choose the best codepath

There is multiple issues with this file. GLib have native .ini/.dektop parsing, we don't need to do it here. GIO have asynchronous .ini/desktop directory loading and GTK have native .desktop icon lookup based on the user theme with all the fallback options. So in the end, no shell commands or parsing is necessary. We should depend on GIO, it has some awesome APIs to fix awesome lockup/single thread problems. We should not depend on GTK+, but use it when it's there.

**Comment by blueyed** [comment]
Ok, I'm looking into merging this, and found `awful.util.geticonpath`, used by naughty:
https://github.com/awesomeWM/awesome/blob/master/lib/awful/util.lua.in#L179-L208

It uses a few more icontypes:

>    local icontypes = { 'apps', 'actions',  'categories',  'emblems', 'mimetypes',  'status', 'devices', 'extras', 'places', 'stock' }

The following are not used in this PR. Is there a specific reason for this?

> 'emblems', 'mimetypes', 'extras', 'stock'

I think there should be a central function, used by naughty and the menubar to get the icon paths probably. Agreed?

**Comment by Elv13** [comment]
Applications should be in `apps` and `apps` only

**Comment by blueyed** [comment]
> Applications should be in apps and apps only

Ok. Removed the others: https://github.com/blueyed/awesome/commit/8d7b554bb527af0be227fed6eeedc1f0e9687203

I've adopted this PR into a new branch, and changed some things on top: https://github.com/blueyed/awesome/compare/awesomeWM:master...blueyed:menubar-icons

@salorium @Elv13 
Ok to merge it like that for now?

**Comment by blueyed** [comment]
@salorium 
Your last commit is missing (https://github.com/salorium/awesome/commit/fecd1404e8c07b7644f048eea4015919dae70be8); I've done it in a different way already, and that seems to make it easier to refactor it with naughty's method later probably.

**Comment by Elv13** [comment]
Feel free to merge it when you feel its ready

**Comment by blueyed** [comment]
Committed in https://github.com/awesomeWM/awesome/commit/2d54ee271dee96fa127ab982289cb5feac06abbe.

@salorium 
Thanks for your contribution! :+1:

**Comment by salorium** [comment]
Svg is supported by awesome ?

**Comment by blueyed** [comment]
It appeared to me:

> https://github.com/awesomeWM/awesome/commit/dc41e62f9a58faf723c3eefba729b1e2362f2b2f

**Pull request closed without merging**